platform/darwin/src/NSExpression+MGLAdditions.h

```diff
@@ -42,6 +90,133 @@ NS_ASSUME_NONNULL_BEGIN
+ colors for each condition.
+ 
+ @param conditionPredicate The predicate to get evaluated.
+ @param trueColor The color for conditions equal to true.
+ @param falseColor The color value if the condition is equal to false.
+ */
++ (instancetype)mgl_expressionForConditional:(nonnull NSString *)conditionPredicate trueColor:(nonnull MGLColor*)trueColor falseColor:(nonnull MGLColor *)falseColor;
+
+/**
+ Returns a conditional function expression specifying the string predicate, and
+ values for each condition.
+ 
+ @param conditionPredicate The predicate to get evaluated.
+ @param trueValue The value for conditions equal to true.
+ @param falseValue The value for conditions equal to false.
+ */
++ (instancetype)mgl_expressionForConditional:(nonnull NSString *)conditionPredicate trueValue:(nonnull NSValue *)trueValue falseValue:(nonnull NSValue *)falseValue;
+
+/**
+ Returns a conditional function expression specifying the string predicate, and
+ expressions for each condition.
+ 
+ @param conditionPredicate The predicate to get evaluated.
+ @param trueExpression The expression for conditions equal to true.
+ @param falseExpression The expression for conditions equal to false.
+ */
++ (instancetype)mgl_expressionForConditional:(nonnull NSString *)conditionPredicate trueExpression:(nonnull NSExpression *)trueExpression falseExpresssion:(nonnull NSExpression *)falseExpression;
+
+/**
+ Returns a step function expression specifying the function operator, default value
+ and stops.
+ 
+ @param function The operator type in which this expression is applied.
+ @param value The default value can be boolean or numeric.
+ @param stops The stops dictionay must be numeric literals in strictly ascending order.
+ */
++ (instancetype)mgl_expressionForStepFunction:(nonnull MGLExpressionStyleFunction)function defaultValue:(nonnull NSValue *)value stops:(nonnull NS_DICTIONARY_OF(NSNumber *, id) *)stops;
+
+/**
+ Returns a step function expression specifying the function operator, default color
+ and stops.
+ 
+ @param function The operator type in which this expression is applied.
+ @param color The default color.
+ @param stops The stops dictionay must be numeric literals in strictly ascending order.
+ */
++ (instancetype)mgl_expressionForStepFunction:(nonnull MGLExpressionStyleFunction)function defaultColor:(nonnull MGLColor *)color stops:(nonnull NS_DICTIONARY_OF(NSNumber *, id) *)stops;
+
+/**
+ Returns a step function expression specifying the function operator, default expression
+ and stops.
+ 
+ @param operatorExpression The operator expression.
+ @param expression The expression which could be a constant or function expression.
+ @param stops The stops dictionay must be numeric literals in strictly ascending order.
+ */
++ (instancetype)mgl_expressionForStepFunction:(nonnull NSExpression*)operatorExpression defaultExpression:(nonnull NSExpression *)expression stops:(nonnull NSExpression*)stops;
+
```

Comment:
It is a little confusing to pair `[NSExpression expressionForConstantValue:]` with a stops dictionary. I would not expect that to be paired with a stops dictionary, especially with exponential interpolation mode.

Also, capturing that in Swift, `steps: NSExpression(format: "\(stops)")` does not seem to work. `steps: NSExpression(format: "%@", stops)` does work.

Reply:
This is expected. Unlike with `String(format:)`, the string representation of an object is not necessarily the same as the literal representation that NSExpression expects in a format string. `NSExpression(format: "\(array)")` would also likely fail, because the literal representation of an array would look like `{1, 2, 3, 4}`, not:

```
(
    1,
    2,
    3,
    4
)
```

In this case, there is no literal representation for a dictionary.